### ISSUE TITLE
Bug or new feature? tbl_df does not have same behavior as data.frame when using `[`

### ISSUE THREAD

**Issue by kapsmane** [hadley/dplyr#1202]
in R 3.0.3 and dplyr 0.2, if i do the following:
 xx=data.frame(a=c(1,2,3),b=c('sdf','sdfsd',"sdafsdf"),stringsAsFactors = F)
 xx_tbl=tbl_df(xx)
class(xx_tbl[,"a"])
output is : [1] "numeric"

However, in R 3.1.3 and dplyr 0.4.1, if i do the similar, 
 class(xx_tbl$a)
[1] "numeric"
 class(xx_tbl[,"a"])
[1] "tbl_df"     "data.frame"

This creates trouble since sometimes i need use a variable to reference different columns, and assign the value. 
For example:
 mean(xx_tbl[,'a'])
[1] NA
Warning message:
In mean.default(xx_tbl[, "a"]) :
  argument is not numeric or logical: returning NA

In other words, tbl_df no longer has the same behavior as data.frame. Notice that the data.frame itself can be referenced this way:
 xx[,"a"]
[1] 1 2 3
 class(xx[,"a"])
[1] "numeric"
 
Is this an expected feature? Thank you.

Best,
Ricky

**Comment by hadley** [comment]
This is expected. If you want to extract a single column, use `[[`

**Comment by kapsmane** [comment]
thank you Hadley, this is very helpful.
